lib/uknofault/tests/test_nofault.c

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/nofault.h>
+#include <uk/plat/paging.h>
+#include <uk/assert.h>
+#include <uk/test.h>
+#include <uk/essentials.h>
+
+#define TEST_MAP_BASE	0xC0000000 /* 3 GiB - must not be used otherwise */
+#define TEST_MAP_END	(TEST_MAP_BASE + PAGE_SIZE)
+
+UK_TESTCASE(uknofault, test_nofault_probe_noaccess)
+{
+	__sz len;
+
+	len = uk_nofault_probe_r(TEST_MAP_BASE, 1);
+	UK_TEST_EXPECT_SNUM_EQ(len, 0);
+
+	len = uk_nofault_probe_r(TEST_MAP_BASE, PAGE_SIZE);
+	UK_TEST_EXPECT_SNUM_EQ(len, 0);
+}
+
+/**
+ * Tests if we can successfully probe for read access.
+ */
+UK_TESTCASE(uknofault, test_nofault_probe_r)
+{
+	struct uk_pagetable *pt = ukplat_pt_get_active();
+	__sz len;
+	int rc;
+
+	rc = ukplat_page_map(pt, TEST_MAP_BASE, __PADDR_ANY, 1,
+			     PAGE_ATTR_PROT_READ, 0);
+	UK_ASSERT(rc == 0);
```

Comment:
This is why I am not happy with the current uktest API: `UK_ASSERT` isn't optimal because you can disable assertions and then this check will just disappear, ignoring any failures silently.
But `UK_TEST_EXPECT_ZERO` also won't do the job, because that won't fail the test immediately. (google test had two macro variants `EXPECT` and `ASSERT` to distinguish between fatal and non-fatal checks...)

Maybe convert this to a `if (cond) { UK_CRASH(...); }` or `UK_TEST_EXPECT_ZERO(...); if (...} { UK_CRASH(...); }`

Reply:
Yep, makes sense. Until we add something like you suggested, I am using the `if (cond) { UK_CRASH(); }` solution.

Reply:
@mschlumpp could you put together an issue describing the the limitations you see with `EXPECT` and `ASSERT` so we can keep track of this?

Reply:
Yes, will do